src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1712,15 +1710,26 @@ private final void putCharsAt(int index, CharSequence s, int off, int end) {
         }
     }
 
-    private void putStringAt(int index, String str, int off, int end) {
-        if (getCoder() != str.coder()) {
+    private void inflateIfNeededFor(String input) {
+        if (COMPACT_STRINGS && (coder != input.coder())) {
+            inflate();
+        }
+    }
+
+    private void inflateIfNeededFor(AbstractStringBuilder input) {
+        if (COMPACT_STRINGS && (coder != input.getCoder())) {
             inflate();
         }
+    }
+
+    private void putStringAt(int index, String str, int off, int end) {
+        inflateIfNeededFor(str);
         str.getBytes(value, off, index, coder, end - off);
     }
 
     private void putStringAt(int index, String str) {
```

Comment:
Can we replace all the calls to this method with calls to previous method as `putStringAt(index, str, 0, str.length())`?

Reply:
No, I don't think so. The only use of this I can find is at line 1298 which effectively adds a substring: `putStringAt(dstOffset, (String) s, start, end);`

Reply:
What about lines 582, 1003 and 1175? E.g. 582
```java
public AbstractStringBuilder append(String str) {
    if (str == null) {
        return appendNull();
    }
    int len = str.length();
    ensureCapacityInternal(count + len);
    putStringAt(count, str);             // couldn't it be putStringAt(count, str, 0, len);
    count += len;
    return this;
}
```
Doing this here and in other places allows  to rid `private void putStringAt(int index, String str)` completely and reduce one nested method call, right?

Reply:
I think you've got the wrong idea: We _want_ to use `putStringAt(int, String)` now since it can call into `String.getBytes(String, int, byte)`, which has a simpler and more efficient implementation than `String.getBytes(String, int, int, byte, int)`, since it avoids a couple of `<< coder` operations. This makes up for most of the improvement between my initial and the current version of this patch. 

(There's also no nested delegation from `putStringAt(int, String)` to `putStringAt(int, String, int, int)` in my proposed patch.)

Reply:
You are right, I've applied your patch locally incompletely and misunderstood it, my bad :)